api/src/main/java/com/netflix/iceberg/io/FileIO.java

```diff
@@ -47,4 +46,30 @@
    * Delete the file at the given path.
    */
   void deleteFile(String path);
+
+  /**
+   * Get an {@link InputFile} to get the bytes for this table's metadata file with the given name.
+   */
+  InputFile readMetadataFile(String fileName);
+
+  /**
+   * Get an {@link OutputFile} to write bytes for a new table metadata file with the given name.
+   */
+  OutputFile newMetadataFile(String fileName);
+
+  /**
+   * Get an {@link OutputFile} for writing bytes to a new data file for this table.
+   * <p>
+   * The partition values of the rows in this file may be used to derive the final location of
+   * the file.
+   */
+  OutputFile newPartitionedDataFile(
```

Comment:
I was thinking about this a little bit more, and in contrast to what was discussed on the ticket, I propose creating the `OutputFile` instance directly instead of just returning the path alone.

This comes up in the case where the file doesn't exist yet, which is always going to be the case for Spark (Spark is always creating new data files). In such a case, it is desirable for the plugin to not necessarily only know what the path is, but also to perhaps create the path itself, oftentimes by asking for some new entry in an external system that is a reference to this file we're about to write to.

In other words, the semantics of this API are that we're not only asking for the location of the data file, but we're also asking to create the data file as part of this action. In practice most of the implementations should just derive this as `newOutputFile(getPath(args))`. Nevertheless I think this version of the API is more rich. Thoughts?

Reply:
Sounds reasonable to me, as long as this doesn't cause odd practices like generating an OutputFile, converting it to a location, then later getting an OutputFile for it. I don't think that would happen though, so this is probably saving a call.

Reply:
We're often using the location of the `OutputFile` object, but we shouldn't be converting twice anywhere.